File: backend/modules/tour_finder.py
"""
Hovedmodul: Finner og rangerer splitboard-turer basert på
vær, skredvarsel, brukerens posisjon og preferanser.
"""

from math import radians, sin, cos, sqrt, atan2
from .known_tours import get_tours_by_region, TOURS
from .weather import get_forecast, evaluate_weather
from .avalanche import get_avalanche_warning, evaluate_avalanche_danger
from .routing import get_drive_time
from .sunrise import get_daylight
from .regobs import get_nearby_observations


def haversine_km(lat1, lon1, lat2, lon2):
    """Avstand mellom to koordinater i km."""
    R = 6371
    dlat = radians(lat2 - lat1)
    dlon = radians(lon2 - lon1)
    a = sin(dlat / 2) ** 2 + cos(radians(lat1)) * cos(radians(lat2)) * sin(dlon / 2) ** 2
    return R * 2 * atan2(sqrt(a), sqrt(1 - a))


def estimate_drive_time_hours(distance_km):
    """
    Grov estimering av kjøretid basert på avstand.
    Fjellveier i Norge: ca 50–60 km/t gjennomsnitt.
    """
    return distance_km / 55


def find_tours(
    user_lat: float = None,
    user_lon: float = None,
    max_drive_hours: float = 4.0,
    target_date: str = None,
    difficulty: str = None,
    region: str = None,
    min_slope: float = 15,
    max_slope: float = 30,
) -> list:
    """
    Finn og ranger turer basert på alle kriterier.
    Returnerer sortert liste med turforslag.
    """
    # Hent alle turer, eventuelt filtrert på region
    tours = get_tours_by_region(region) if region else TOURS

    results = []

    # Cache for API-kall per region
    weather_cache = {}
    avalanche_cache = {}
    sunrise_cache = {}
    regobs_cache = {}

    for tour in tours:
        # --- Avstandsfilter ---
        if user_lat and user_lon:
            # Prøv OSRM først, fallback til haversine
            osrm_result = get_drive_time(
                user_lat, user_lon,
                tour["start"]["lat"], tour["start"]["lon"]
            )
            if osrm_result:
                dist, drive_hours = osrm_result
                drive_source = "osrm"
            else:
                dist = haversine_km(
                    user_lat, user_lon,
                    tour["start"]["lat"], tour["start"]["lon"]
                )
                drive_hours = estimate_drive_time_hours(dist)
                drive_source = "estimate"
            if drive_hours > max_drive_hours:
                continue
        else:
            dist = None
            drive_hours = None
            drive_source = None

        # --- Helningsfilter ---
        slope = tour.get("slope_avg_deg", 25)
        if slope < min_slope or slope > max_slope:
            continue

        # --- Vanskelighetsfilter ---
        if difficulty and difficulty.lower() not in tour.get("difficulty", "").lower():
            continue

        # --- Hent vær (cachet per koordinat-område) ---
        weather_key = f"{round(tour['summit']['lat'], 1)}_{round(tour['summit']['lon'], 1)}"
        if weather_key not in weather_cache:
            try:
                forecasts = get_forecast(tour["summit"]["lat"], tour["summit"]["lon"])
            except Exception as e:
                print(f"Vær-API feilet for {tour['name']}: {e}")
                forecasts = None
            weather_cache[weather_key] = forecasts

        weather_eval = evaluate_weather(weather_cache[weather_key], target_date)

        # --- Hent skredvarsel (cachet per region) ---
        region_id = tour["varsom_region_id"]
        if region_id not in avalanche_cache:
            try:
                warnings = get_avalanche_warning(region_id)
            except Exception as e:
                print(f"Skred-API feilet for region {region_id}: {e}")
                warnings = None
            avalanche_cache[region_id] = warnings

        aval_eval = evaluate_avalanche_danger(avalanche_cache[region_id], target_date)

        # --- Hent soldata (cachet per koordinat-område) ---
        sunrise_key = f"{round(tour['summit']['lat'], 1)}_{round(tour['summit']['lon'], 1)}"
        if sunrise_key not in sunrise_cache:
            try:
                sunrise_cache[sunrise_key] = get_daylight(
                    tour["summit"]["lat"], tour["summit"]["lon"], target_date
                )
            except Exception as e:
                print(f"Sunrise-API feilet for {tour['name']}: {e}")
                sunrise_cache[sunrise_key] = None

        sunrise_data = sunrise_cache[sunrise_key]

        # --- Hent RegObs-observasjoner (cachet per region-område) ---
        regobs_key = f"{round(tour['summit']['lat'], 0)}_{round(tour['summit']['lon'], 0)}"
        if regobs_key not in regobs_cache:
            try:
                regobs_cache[regobs_key] = get_nearby_observations(
                    tour["summit"]["lat"], tour["summit"]["lon"]
                )
            except Exception as e:
                print(f"RegObs-API feilet for {tour['name']}: {e}")
                regobs_cache[regobs_key] = []

        nearby_obs = regobs_cache[regobs_key]

        # --- Total score ---
        # Vekting: skredfare er viktigst, deretter vær, deretter avstand
        aval_score = aval_eval["score"]
        weather_score = weather_eval["score"]
        distance_score = 100 - (drive_hours / max_drive_hours * 30) if drive_hours else 70

        total_score = (
            aval_score * 0.50       # Skredfare veier tyngst
            + weather_score * 0.35  # Vær er viktig
            + distance_score * 0.15 # Avstand er bonus
        )

        # Sikkerhetsregel: faregrad >= 4 → tur frarådes uansett
        if aval_eval.get("danger_level", 0) >= 4:
            total_score = min(total_score, 10)

        results.append({
            "tour": tour,
            "total_score": round(total_score, 1),
            "weather": weather_eval,
            "avalanche": aval_eval,
            "distance_km": round(dist, 1) if dist else None,
            "drive_hours": round(drive_hours, 1) if drive_hours else None,
            "drive_source": drive_source,
            "sunrise": sunrise_data,
            "observations": nearby_obs[:3],
        })

    # Sorter etter total score (best først)
    results.sort(key=lambda x: x["total_score"], reverse=True)

    return results

File: js/tour-finder.js
// ── Toppturfinner — Tour Finder ──
// Filtrering, scoring og rangering av turer.
// Bruker TOURS fra tours-data.js og API-funksjoner fra api.js.

function haversineKm(lat1, lon1, lat2, lon2) {
    const R = 6371;
    const dLat = (lat2 - lat1) * Math.PI / 180;
    const dLon = (lon2 - lon1) * Math.PI / 180;
    const a = Math.sin(dLat / 2) ** 2
        + Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180)
        * Math.sin(dLon / 2) ** 2;
    return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function estimateDriveHours(distanceKm) {
    return distanceKm / 55;
}

/**
 * Finn og ranger turer basert på alle kriterier.
 * @param {Object} opts
 * @param {number|null} opts.userLat
 * @param {number|null} opts.userLon
 * @param {number} opts.maxDriveHours
 * @param {string|null} opts.targetDate - YYYY-MM-DD
 * @param {string|null} opts.difficulty
 * @param {string|null} opts.region
 * @param {number} opts.minSlope
 * @param {number} opts.maxSlope
 * @param {function|null} opts.onProgress - callback(message) for progress updates
 * @returns {Promise<Array>}
 */
async function findTours({
    userLat = null,
    userLon = null,
    maxDriveHours = 4.0,
    targetDate = null,
    difficulty = null,
    region = null,
    minSlope = 15,
    maxSlope = 30,
    onProgress = null,
} = {}) {
    // Filtrer turer
    let tours = TOURS;

    if (region) {
        tours = tours.filter(t => t.region.toLowerCase() === region.toLowerCase());
    }

    // Helningsfilter
    tours = tours.filter(t => {
        const slope = t.slope_avg_deg || 25;
        return slope >= minSlope && slope <= maxSlope;
    });

    // Vanskelighetsfilter
    if (difficulty) {
        tours = tours.filter(t =>
            (t.difficulty || '').toLowerCase().includes(difficulty.toLowerCase())
        );
    }

    if (onProgress) onProgress(`Filtrert: ${tours.length} turer funnet`);

    // ── Steg 1: Beregn kjøretid for alle turer (parallelt) ──
    if (onProgress) onProgress('Beregner kjøretid...');

    let toursWithDrive = [];

    if (userLat != null && userLon != null) {
        const drivePromises = tours.map(async (tour) => {
            const osrmResult = await fetchDriveTime(
                userLat, userLon,
                tour.start.lat, tour.start.lon
            );

            let dist, driveHours, driveSource;
            if (osrmResult) {
                dist = osrmResult.distance_km;
                driveHours = osrmResult.duration_hours;
                driveSource = 'osrm';
            } else {
                dist = haversineKm(userLat, userLon, tour.start.lat, tour.start.lon);
                driveHours = estimateDriveHours(dist);
                driveSource = 'estimate';
            }

            if (driveHours > maxDriveHours) return null;

            return { tour, dist, driveHours, driveSource };
        });

        const driveResults = await Promise.all(drivePromises);
        toursWithDrive = driveResults.filter(r => r !== null);
    } else {
        toursWithDrive = tours.map(tour => ({
            tour,
            dist: null,
            driveHours: null,
            driveSource: null,
        }));
    }

    if (onProgress) onProgress(`${toursWithDrive.length} turer innenfor kjøreavstand`);

    // ── Steg 2: Hent skredvarsel (cachet per region, parallelt) ──
    if (onProgress) onProgress('Henter skredvarsel...');

    const uniqueRegions = [...new Set(toursWithDrive.map(r => r.tour.varsom_region_id))];
    const avalancheResults = {};

    await Promise.all(uniqueRegions.map(async (regionId) => {
        const cacheKey = `aval_${regionId}`;
        if (_avalancheCache.has(cacheKey)) {
            avalancheResults[regionId] = _avalancheCache.get(cacheKey);
        } else {
            try {
                const warnings = await fetchAvalanche(regionId);
                _avalancheCache.set(cacheKey, warnings);
                avalancheResults[regionId] = warnings;
            } catch (e) {
                console.warn(`Skred-API feilet for region ${regionId}: ${e.message}`);
                avalancheResults[regionId] = null;
            }
        }
    }));

    // ── Steg 3: Hent vær, soldata og RegObs (parallelt, cachet per område) ──
    if (onProgress) onProgress('Henter vær og soldata...');

    const results = [];

    // Grupper API-kall for å unngå duplikater
    const weatherPromises = new Map();
    const sunrisePromises = new Map();
    const regobsPromises = new Map();

    for (const item of toursWithDrive) {
        const t = item.tour;

        // Vær: cache per 0.1° koordinat
        const weatherKey = `${t.summit.lat.toFixed(1)}_${t.summit.lon.toFixed(1)}`;
        if (!weatherPromises.has(weatherKey)) {
            if (_weatherCache.has(weatherKey)) {
                weatherPromises.set(weatherKey, Promise.resolve(_weatherCache.get(weatherKey)));
            } else {
                weatherPromises.set(weatherKey, fetchWeather(t.summit.lat, t.summit.lon).then(f => {
                    _weatherCache.set(weatherKey, f);
                    return f;
                }));
            }
        }

        // Sunrise: cache per 0.2° koordinat
        const sunKey = `${(Math.round(t.summit.lat * 5) / 5).toFixed(1)}_${(Math.round(t.summit.lon * 5) / 5).toFixed(1)}_${targetDate || 'default'}`;
        if (!sunrisePromises.has(sunKey)) {
            if (_sunriseCache.has(sunKey)) {
                sunrisePromises.set(sunKey, Promise.resolve(_sunriseCache.get(sunKey)));
            } else {
                sunrisePromises.set(sunKey, fetchDaylight(t.summit.lat, t.summit.lon, targetDate).then(s => {
                    _sunriseCache.set(sunKey, s);
                    return s;
                }));
            }
        }

        // RegObs: cache per 1° koordinat
        const regKey = `${Math.round(t.summit.lat)}_${Math.round(t.summit.lon)}`;
        if (!regobsPromises.has(regKey)) {
            if (_regobsCache.has(regKey)) {
                regobsPromises.set(regKey, Promise.resolve(_regobsCache.get(regKey)));
            } else {
                regobsPromises.set(regKey, fetchRegObs(t.summit.lat, t.summit.lon).then(o => {
                    _regobsCache.set(regKey, o);
                    return o;
                }));
            }
        }
    }

    // Vent på alle API-kall
    await Promise.all([
        ...weatherPromises.values(),
        ...sunrisePromises.values(),
        ...regobsPromises.values(),
    ]);

    if (onProgress) onProgress('Beregner score...');

    // ── Steg 4: Bygg resultater ──
    for (const item of toursWithDrive) {
        const t = item.tour;

        // Hent cached data
        const weatherKey = `${t.summit.lat.toFixed(1)}_${t.summit.lon.toFixed(1)}`;
        const forecasts = _weatherCache.get(weatherKey) || null;
        const weatherEval = evaluateWeather(forecasts, targetDate);

        const avalWarnings = avalancheResults[t.varsom_region_id];
        const avalEval = evaluateAvalanche(avalWarnings, targetDate);

        const sunKey = `${(Math.round(t.summit.lat * 5) / 5).toFixed(1)}_${(Math.round(t.summit.lon * 5) / 5).toFixed(1)}_${targetDate || 'default'}`;
        const sunriseData = _sunriseCache.get(sunKey) || null;

        const regKey = `${Math.round(t.summit.lat)}_${Math.round(t.summit.lon)}`;
        const nearbyObs = _regobsCache.get(regKey) || [];

        // Total score
        const avalScore = avalEval.score;
        const weatherScore = weatherEval.score;
        const distanceScore = item.driveHours != null
            ? 100 - (item.driveHours / maxDriveHours * 30)
            : 70;

        let totalScore = (
            avalScore * 0.50 +
            weatherScore * 0.35 +
            distanceScore * 0.15
        );

        // Sikkerhetsregel: faregrad >= 4 → score maks 10
        if ((avalEval.danger_level || 0) >= 4) {
            totalScore = Math.min(totalScore, 10);
        }

        results.push({
            tour: t,
            total_score: Math.round(totalScore * 10) / 10,
            weather: weatherEval,
            avalanche: avalEval,
            distance_km: item.dist != null ? Math.round(item.dist * 10) / 10 : null,
            drive_hours: item.driveHours != null ? Math.round(item.driveHours * 10) / 10 : null,
            drive_source: item.driveSource,
            sunrise: sunriseData,
            observations: nearbyObs.slice(0, 3),
        });
    }

    // Sorter etter total score (best først)
    results.sort((a, b) => b.total_score - a.total_score);

    if (onProgress) onProgress(`Ferdig! ${results.length} turer rangert`);

    return results;
}
